File: 1361. Validate Binary Tree Nodes.py
# 1361. Validate Binary Tree Nodes

class Solution:
    def validateBinaryTreeNodes(self, n: int, leftChild: List[int], rightChild: List[int]) -> bool:
        # bfs solution
        children = set(leftChild + rightChild)
        children.discard(-1)

        if len(children) != n-1:
            return False
        
        root = None
        for i in range(n):
            if i not in children:
                root = i
                break
        
        visit = set()
        q = deque([root])
        visit.add(root)
        
        while q:
            node = q.popleft()
            l = leftChild[node]
            r = rightChild[node]
            if (l in visit) or (r in visit):
                return False

            if l != -1:
                visit.add(l)
                q.append(l)
            if r != -1:
                visit.add(r)
                q.append(r)
        return True if len(visit) == n else False
    

        # dfs solution
        # no cycle, all nodes should be connected, 1 root

        # child = set(leftChild + rightChild)
        # child.discard(-1)
        # if len(child) != n-1:
        #     return False
        
        # root = -1
        # for i in range(n):
        #     if i not in child:
        #         root = i
        #         break
        
        # #find cycle
        # visit = set()
        # def dfs(node):
        #     if node == -1:
        #         return True
        #     if node in visit:
        #         return False

        #     visit.add(node)

        #     if leftChild[node] == -1 and rightChild[node] == -1:
        #         return True

        #     return dfs(leftChild[node]) and dfs(rightChild[node])
            
        
        # return dfs(root) and len(visit) == n

File: 1361. Validate Binary Tree Nodes.cpp
// 1361. Validate Binary Tree Nodes

#include <vector>
#include <queue>
using namespace std;


class Solution {
    public:
        bool validateBinaryTreeNodes(int n, vector<int>& leftChild, vector<int>& rightChild) {
            vector<int> roots(n, -1);
    
            for (int i = 0; i < n; i++) {
                if (leftChild[i] != -1){
                    if (roots[leftChild[i]] != -1) {
                        return false;
                    }
                    else {
                        roots[leftChild[i]] = i;
                    }
                }
            }
    
            for (int i = 0; i < n; i++) {
                if (rightChild[i] != -1){
                    if (roots[rightChild[i]] != -1) {
                        return false;
                    }
                    else {
                        roots[rightChild[i]] = i;
                    }
                }
            }
    
            int root = -1;
            int root_count = 0;
            for (int i = 0; i < n; i++) {
                if (roots[i] == -1) {
                    root = i;
                    root_count++;
                }
            }
    
            if (root == -1 || root_count > 1) return false;
            
            int count = 0;
            vector<bool> visit(n, false);
            queue<int> q;
            q.push(root);
            while (!q.empty()) {
                int node = q.front();
                q.pop();
                if (visit[node] == true) {return false;}
                visit[node] = true;
                count++;
                if (leftChild[node] != -1) {
                    q.push(leftChild[node]);
                }
                if (rightChild[node] != -1) {
                    q.push(rightChild[node]);
                }
            }
            return n == count;
    
        }
    };